File: leetCode/Python/242-validAnagram.py
# Given two strings s and t, write a function 
# to determine if t is an anagram of s.

# For example,
# s = "anagram", t = "nagaram", return true.
# s = "rat", t = "car", return false.

# Note:
# You may assume the string contains only lowercase alphabets.


class Solution(object):
    def isAnagram(self, s, t):
        '''
        Time:  O(nlogn)
        Space: O(1)
        '''
        return sorted(s) == sorted(t)

class Solution(object):
    def isAnagram(self, s, t):
        '''
        Time:  O(n)
        Space: O(n)
        '''
        if len(s) != len(t):
            return False
        tdict = {}; sdict = {}
        for char in s:
            sdict[char] = sdict.get(char, 0) + 1
        for char in t:
            tdict[char] = tdict.get(char, 0) + 1
        for char in s:
            if sdict.get(char) != tdict.get(char):
                return False
        return True

File: leetCode/C++/242-validAnagram.cc
#include <algorithm>
#include <iostream>
#include <unordered_map>

using namespace std;

class Solution1 {
public:
    bool isAnagram(string s, string t) {
        /*
            Time:  O(nlogn)
            Space: O(1)
        */
         if (s.length() != t.length())
             return false;
         sort(s.begin(), s.end());
         sort(t.begin(), t.end());
         return s == t;
    }
};

class Solution {
public:
    bool isAnagram(string s, string t) {
        /*
            Time:  O(n)
            Space: O(n)
        */
         unordered_map<char, int> count;
         if (s.length() != t.length())
             return false;
         for (const auto &c: s)
            ++count[tolower(c)];
         for (const auto &c: t) {
            --count[tolower(c)];
            if (count[tolower(c)] < 0)
                return false;
        }
         return true;
    }
};


int main(void) {
    Solution s;
    string test1 = "bdac", test2 = "abcd";
    cout << s.isAnagram(test1, test2) << endl;
    test1 = "car", test2 = "rat";
    cout << s.isAnagram(test1, test2) << endl;
    return 0;
};